notes/java.md

```diff
@@ -164,4 +164,47 @@ One thing that could be added here (or somewhere else) is, that there should be
         }
         return checkedFunDefs;
 
+```
+
+Now we will give an oversimplified sketch of the interpreter. All the functions below should be changed to solve lab 2 and the interface `Value` and its implementations (e.g. `Value.Void`) have to be defined.
+
+```java
+    void interpret(TypeChecker.AnnotatedProgram p) {
+        defs = p.defs();
+        if (!defs.containsKey("main")) {
+            throw new RuntimeException("Main function not found");
+        }
+        runFunction(defs.get("main")); // TODO: deal with arguments...
+    }
+
+    // TODO: Extends to functions with arguments
+    Value runFunction(TypeChecker.TypedFunDef f) {
+        var env = new Environment();
+        for (var s : f.stms()) {
+            run(s, env);
+        }
+        return new Value.Void();
+    }
+
+    // TODO: Instead of 'env' there should be a stack or list of environments
+    void run(Statement s, Environment env) {
+        switch (s) {
+            case Statement.Decl decl -> env.setVar(decl.name(), decl.type());
+            case Statement.Expr expr -> evalExpr(expr.typedExpr(), env);
+        };
+    }
+
+```
+
+Depending on how `return` statements are implemented, it might be reasonable to return more information here than just `Value`s. For the following evaluation function for expressions, it should however be enough to return only a `Value`:
+
+```java
+    Value evalExpr(TypedExpr expr, Environment env) {
```

Comment:
I think the `Environment` should be a field of the interpreter class that is then updated by `evalExpr`.

In the current formulation, `evalExpr` returns a `Value` but also has a side effect on one of its arguments (that is not returned).
This is unfortunately not in line with the lecture materials, which advocate 
1. either to return also the updated `Environment` (manual threading)
2. or hide this under a State monad (which in Java would be a field of the class which is then mutated).

I think the style you are proposing leads to confusion when matched against the threading style.
We should not propose yet another method...